File: internal/cli/serve.go
package cli

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/CrowdStrike/codestrike/internal/api"
	"github.com/CrowdStrike/codestrike/internal/config"
	"github.com/CrowdStrike/codestrike/internal/setup"
	"github.com/CrowdStrike/codestrike/internal/setup/logger"
	"github.com/emicklei/go-restful/v3"
	"github.com/joho/godotenv"
	"github.com/rs/zerolog"
	"github.com/spf13/cobra"
)

func newServeCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "serve",
		Short: "Run CodeStrike as a persistent HTTP service",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			_ = godotenv.Load()

			envCfg := setup.LoadConfig()
			log := logger.New(envCfg.LogLevel)

			appConfig, err := getAppConfig(cmd)
			if err != nil {
				return err
			}

			if envCfg.GitHubToken == "" {
				return fmt.Errorf("GITHUB_TOKEN environment variable is required")
			}

			llmClient, err := setup.CreateLLMClient(cmd.Context(), envCfg)
			if err != nil {
				return fmt.Errorf("creating LLM client: %w", err)
			}

			addr, _ := cmd.Flags().GetString("addr")
			if addr == "" {
				addr = ":" + envCfg.ListenHttpPort
			}

			handler := api.NewHandler(llmClient, appConfig, envCfg.GitHubToken, &log)
			container := restful.NewContainer()
			api.RegisterRoutes(container, handler)

			srv := &http.Server{
				Addr:              addr,
				Handler:           container,
				ReadHeaderTimeout: 10 * time.Second,
			}

			return listenAndShutdown(cmd.Context(), srv, &log)
		},
	}

	cmd.Flags().String("addr", "", "Bind address (default :8080, or PORT env var)")

	return cmd
}

func getAppConfig(cmd *cobra.Command) (*config.Config, error) {
	configPath, err := cmd.Flags().GetString("config")
	if err != nil {
		return nil, fmt.Errorf("reading --config flag: %w", err)
	}

	resolvedPath, err := config.ResolvePath(configPath)
	if err != nil {
		return nil, fmt.Errorf("resolving config path: %w", err)
	}

	appConfig, err := config.Load(resolvedPath)
	if err != nil {
		return nil, fmt.Errorf("loading config: %w", err)
	}
	return appConfig, nil
}

func listenAndShutdown(ctx context.Context, srv *http.Server, logger *zerolog.Logger) error {
	errCh := make(chan error, 1)

	go func() {
		logger.Info().Str("addr", srv.Addr).Msg("server listening")
		errCh <- srv.ListenAndServe()
	}()

	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)

	select {
	case err := <-errCh:
		if err != nil && err != http.ErrServerClosed {
			return fmt.Errorf("server error: %w", err)
		}
	case sig := <-sigCh:
		logger.Info().Str("signal", sig.String()).Msg("shutting down")
	case <-ctx.Done():
		logger.Info().Msg("context cancelled, shutting down")
	}
	shutdownCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	return srv.Shutdown(shutdownCtx)
}

File: src/cli/serve.ts
import { Command } from 'commander';
import { config as loadDotenv } from 'dotenv';
import { load as loadConfig } from '../config/config.js';
import { resolvePath } from '../config/path.js';
import { loadEnvConfig, buildLLMClient } from '../setup.js';
import { createLogger } from '../logger.js';
import { createServer } from '../api/routes.js';

export function serveCommand(): Command {
  const cmd = new Command('serve')
    .description('Run CodeStrike as a persistent HTTP service')
    .option('--addr <address>', 'Bind address (default :8000, or PORT env var)')
    .action(async (opts: { addr?: string }) => {
      loadDotenv();

      const envCfg = loadEnvConfig();
      const logger = createLogger(envCfg.logLevel);

      const configFlag = cmd.parent?.getOptionValue('config') as string | undefined;
      const configPath = resolvePath(configFlag);
      const appConfig = loadConfig(configPath);

      if (!envCfg.gitHubToken && !envCfg.bitbucketToken) {
        throw new Error('GITHUB_TOKEN or BITBUCKET_TOKEN environment variable is required');
      }

      const llmClient = buildLLMClient(envCfg);

      let addr = opts.addr ?? '';
      if (!addr) addr = `:${envCfg.listenHttpPort}`;

      const [host, portStr] = parseAddr(addr);
      const port = parseInt(portStr, 10);

      const server = createServer(llmClient, appConfig, envCfg, logger);

      try {
        await server.listen({ port, host: host || '0.0.0.0' });
        logger.info({ addr }, 'server listening');
      } catch (err) {
        logger.error({ err }, 'server error');
        process.exit(1);
      }

      const shutdown = async (signal: string) => {
        logger.info({ signal }, 'shutting down');
        await server.close();
        process.exit(0);
      };

      process.on('SIGINT', () => shutdown('SIGINT'));
      process.on('SIGTERM', () => shutdown('SIGTERM'));
    });

  return cmd;
}

function parseAddr(addr: string): [string, string] {
  if (addr.startsWith(':')) return ['', addr.slice(1)];
  const lastColon = addr.lastIndexOf(':');
  if (lastColon === -1) return ['', addr];
  return [addr.slice(0, lastColon), addr.slice(lastColon + 1)];
}
